[Team#01][BE] 3주차 PR입니다

he2joojo opened the pull request:
## 이번 주 작업 내용 🔑

- 리뷰 적용하여 리팩토링
- Mybatis 스프링부트 사용해 필터 적용한 이슈 불러오기 기능 구현
- jwt 로그인 기능 구현
- jwt 로그아웃 기능 구현(임시로 refreshToken을 db에서 삭제하는 방식으로 구현해두었습니다. 시간이 나면 수정할 계획입니다!)
- Label CRUD 구현
- Milestone CRUD 구현

<img width="711" alt="스크린샷 2023-08-11 오전 11 35 19" src="https://github.com/codesquad-members-2023/issue-tracker-max/assets/121915790/6e0fb020-70c0-4883-989e-d77c2ed155a4">

## To Cooper 👋

- 마이바티스만 써서 구현했는데 테스트코드를 짜기가 어려워서  ```mybatis spring boot starter```와 ```mybatis spring boot starter test```를 사용했습니다.
원래 마이바티스 스프링을 쓰기 싫없는데, 이는 스프링이 빈을 자동적으로 등록하고 ```datasource```도 연결해 주지만, 제대로 이해하고 쓰고 싶어서 가장 기본적인 기능이 있는 Mybatis만 써보려고 했습니다.
스프링 빈에 등록하고 ```config.xml```을 만들어서 동적쿼리 까지는 만들었지만, 테스트 코드를 짜는데 실패해서 조금더 간편한 ```mybatis-spring```을 활용했습니다.
주말이 시간이 되면 다시 마이바티스로 롤백해서 테스트까지 완성해보고 싶습니다.
쿠퍼는 이런 학습방법에 대해서 어떻게 생각하시는지 알고싶습니다. 제가 좋은 방향으로 학습하는지 의견 부탁드려요...
- ```JwtIntegrationTest``` 클래스의 ```reissueAccessToken()``` 메서드에서 만들어둔 커스텀 예외가 발생하여 우선 주석처리해두었습니다. 포스트맨으로 확인했을때는 정상작동하는데, 테스트 시 실패해서 테스트코드를 잘못만든 것인지 코드가 잘못된 것인지 아직 모르겠습니다. 조금 더 학습한뒤 수정해서 다시 올리겠습니다!
- 팀프로젝트가 길어지면서 팀원들이 지쳐가는데 쿠퍼는 팀프로젝트를 하면서 어떻게 멘탈 관리를 하셨나요?

nag0124 committed:
[BE]: 리뷰 적용하여 리팩토링 (#53) (#56)

* refactor: sql문에서 약어 삭제 #53

* refactor: Repository 의존성 변경  (#53)

각 repository에서 datasource를 주입했었는데, jdbctemplate을 주입하게 하고 생성자를 롬북으로 관리하여 깔큼한 코드로 변경했습니다.
test에서도 jdbctemplate을 주입하여 repository를 생성하도록 변경했습니다.

* refactor: ControllerTest, IntegrateionTest 변경 (#53)

- 클래스를 public에서 default로 변경해 다른 클래스에서의 접근을 막았습니다.
- Mock 선언 부분과 Autowired 선언부분 사이의 개행을 추가하여 가독성을 향상했습니다.
- httpResponse의 body를 검증하는 부분을 andExpectAll()로 묶어 명백한 코드로 변경했습니다.
- 필요없는 개행을 제거했습니다.

* refactor: MemberRepositoryTest 의존성 주입 변경 (#53)

DataSource가 아니라 jdbcTemplate를 바로 주입받게 변경했습니다.

* Refactor: 테스트 코드 리팩토링

- import문을 정리했습니다.
- 클래스의 public 접근자를 삭제하여 다른 곳에서의 접근을 제한했습니다.
- given, when, then 주석을 추가하여 가독성을 높였습니다.
- assertSoftly()를 사용하여 assertThat을 하나로 묶어 가독성을 높였습니다.
- 메서드 이름에 test를 삭제했고 이름을 수정하여 메서드의 의미가 드러나게 했습니다.

* chore: main 코드의 import, 빈줄 정리

가독성을 위해 정리했습니다.

---------

Co-authored-by: he2joo <hottestej6@gamil.com>

nag0124 committed:
[BE] issueCRUD 브랜치 dev-be에 머지 (#58)

* feat: 선택한 이슈 상태 변경하기 기능 구현 #42

* test: 선택한 이슈 상태 변경하기 통합 테스트 #42

* refactor: 필드 명 isOpen을 open으로 변경 #42

- lombok @Getter가 필드 isOpen의 get메서드명을 자동으로 isOpen()으로 생성해서 필드 명을 변경하고 IssueStatusRequest에 @JsonProperty("isOpen")을 추가했습니다.

* refactor: 필드 명 isOpen을 open으로 변경 #42

- lombok @Getter가 필드 isOpen의 get메서드명을 자동으로 isOpen()으로 생성해서 필드 명을 변경하고 IssueStatusRequest에 @JsonProperty("isOpen")을 추가했습니다.

* feat: 이슈 작성 기능 추가 #42

- IssueRepository에서 issue테이블에 데이터 추가
- MemberRepository에서 issue_label테이블에 데이터 추가
- LabelRepository에서 issue_assignee테이블에 데이터 추가

* feat: 이슈 삭제 기능 추가 #42

- 컬럼 is_deleted = true로 바꾸는 soft delete 기능을 구현했습니다.

* test: 이슈 삭제 기능 통합 테스트 #42

* feat: 이슈 하나의 상태를 변경하는 기능 추가 #42

* test: 이슈 하나의 상태를 변경하는 기능 통합 테스트 #42

* feat: 이슈 내용을 수정하는 기능 추가 #42

* fix: batchUpdate 쿼리 수정 #42

* test: 이슈 수정 기능 통합 테스트 작성 #42

---------

Co-authored-by: he2joo <hottestej6@gamil.com>

nag0124 committed:
fix: 브랜치 머지 후 작동하지 않는 오류 수정

he2joojo committed:
[BE]: JWT 로그인, 로그아웃 기능 구현 (#57) (#62) (#69)

* chore: jwt 의존성 추가 #57

* feat: JwtExceptionType 구현 #57

* feat: JwtRepository 구현 #57

- refresh_token 테이블을 추가하고 refreshToken을 저장하는 찾는 기능을 구현했습니다.

* feat: MemberRepository 구현 #57

- email로 회원을 찾는 기능을 구현했습니다.
- 회원을 저장하는 기능을 구현했습니다.

* feat: JwtService 구현 #57

- JwtProvider로 토큰을 생성하는 기능을 구현했습니다.
- 로그인과 회원가입 시 검증하고 토큰을 반환하는 기능을 구현했습니다.
- refreshToken을 검증하고 accessToken을 재발급하는 기능을 구현했습니다.

* feat: JwtController 구현 #57

- 회원가입, 로그인, 토큰 재발급 api를 구현했습니다.

* feat: JwtAuthorizationFilter 구현 #57

* feat: requestDto에 valid 추가 #57

* feat: GlobalExceptionHandler 구현 #57

* feat: jwt 로그아웃 기능 구현 #57

* refactor: jwt 로그아웃 기능 수정 #57

- claims에 넣어놨던 memberId를 이용해 로그아웃하도록 수정했습니다.

* refactor: repository에서 의존성 주입 방식 변경 #57

- Datasource 대신 NamedParameterJdbcTemplate을 주입받도록 수정했습니다.

* feat: jwt를 사용해 로그인, 로그아웃 기능 구현 #57

* test: 통합테스트 요청 시 토큰 삽입 #57

* test: MemberRepositoryTest에 findByEmail테스트 작성 #57

* refactor: HttpServletRequest을 Service에서 사용하지 않도록 수정 #57

* test: JwtIntegrationTest 작성 #57

---------

Co-authored-by: he2joo <hottestej6@gamil.com>

he2joojo committed:
[BE]: 레이블 CRUD 기능 구현 (#63) (#71)

* feat: 라벨 추가 기능 구현 #63

* test: LabelRepositoryTest 작성 #63

* test: IssueIntegrationTest 토큰 추가 #63

* test: LabelIntegrationTest 작성 #63

* feat: 라벨 상세보기 기능 구현 #63

* test: LabelRepositoryTest에서 findById() 테스트 작성 #63

* test: 라벨 상세보기 통합테스트 작성 #63

* feat: 라벨 수정하기 기능 구현 #63

* test: 라벨 수정하기 LabelRepositoryTest 작성 #63

* test: 라벨 수정하기 LabelIntegrationTest 작성 #63

* feat: 라벨 삭제하기 기능 구현 #63

* test: 라벨 삭제하기 LabelRepositoryTest 작성 #63

* test: 라벨 삭제하기 LabelIntegrationTest 작성 #63

he2joojo committed:
[BE]: 마일스톤 CRUD 기능 구현 (#65) (#73)

* feat: 마일스톤 추가 기능 구현 #65

* test: MilestoneRepositoryTest 작성 #65

* test: MilestoneIntegrationTest 작성 #65

* feat: 마일스톤 수정하기 기능 구현 #65

* test: 마일스톤 수정하기 MilestoneRepositoryTest 작성 #65

* test: 마일스톤 수정하기 MilestoneIntegrationTest 작성 #65

* feat: 마일스톤 삭제하기 기능 구현 #65

* test: 마일스톤 삭제하기 MilestoneRepositoryTest 작성 #65

* test: 마일스톤 삭제하기 MilestoneIntegrationTest 작성 #65

* feat: 마일스톤 닫기/열기 기능 구현 #65

* test: 마일스톤 닫기/열기 기능 MilestoneRepositoryTest 작성 #65

* test: 마일스톤 닫기/열기 기능 MilestoneIntegrationTest 작성 #65

nag0124 committed:
[BE]: be-feature/filter 브랜치 dev-be로 머지 (#77)

* build: 마이바티스 설정 (#59)

build.gradle에 의존성을 추가했고 config와 mapper xml을 만들었습니다.

* feat: MybatisConfig 스프링 configuration 작성 (#59)

SqlSessionFactory를 빈에 등록하여 스프링 실행시 하나만 가지고 있게 했습니다.

* feat: FilteredIssueRepository 생성

쿼리를 제외하고 FilteredIssueRepository를 만들었습니다.

* rename: 패키지 이름 수정 (#53)

패키지 이름이 controller가 아니라 contoller여서 수정했습니다.

* build: 마이바티스 의존성 변경 (#53)

기본적인 마이바티스를 테스트하기 힘들어 마이바티스 스프링 부트로 의존성을 변경했습니다.

* fix: 마이바티스 대신 마이바티스 스프링 부트 스타터 변경

너무 힘듭니다... 이렇게 힘들게 아닌데...
테스트 하기가 어려워 일단 스타터를 사용하여 손쉽게 코드를 짰습니다.
나중에 시간이 있다면 그냥 마이바티스도 해봐야 겠습니다.

* test: findEmpthIssues() 테스트 작성 (#53)

필터링에 맞는 결과가 없으면 빈 리스트를 반환하는 테스트를 작성했습니다.

* feat: IssueService에 FilteredIssueRepository 연결

* feat: 필터 기능 마무리

integration테스트와 controllertest가 없습니다만 일단 기능을 올려야 하기에 올립니다.

* chore: data.sql 수정

프로필 이미지를 다양화 했습니다.

* chore: data.sql 변경

변경을 잘못해서 다시 수정합니다.

* chore: 쓰지 않는 디렉토리 삭제

마이바티스에서 마이바티스 스프링 부트 스타터로 변경하면서 쓰지 않는 파일을 삭제 했습니다.

* chore: 쓰지 않는 test 메서드 삭제

테스트를 위해 만든 test메서드를 삭제했습니다.

* fix: 브랜치 합치는 과정에서 테스트 코드가 돌아가지 않는 부분 수정

바뀐 data.sql과 schema.sql을 testSchema.xml에 덮었습니다.

pbg0205 commented on be/src/main/java/codesquad/kr/gyeonggidoidle/issuetracker/domain/milestone/controller/request/MilestoneRequest.java:
```diff
@@ -0,0 +1,39 @@
+
+    private final String name;
+    private final String description;
+    private final LocalDate dueDate;
```
default constructor 가 없어서 500 에러가 발생하고 있어요. default constructor 와 getter(or setter) 가 필요한 이유를 학습해보시고 이유에 대해서 간단히 코멘트 남겨주세요. 아래 학습할만 내용 첨부드릴게요.

- [[tecoble] @RequestBody vs @ModelAttribute](https://tecoble.techcourse.co.kr/post/2021-05-11-requestbody-modelattribute/)

he2joojo replied:
```final```삭제하고 기본생성자 추가했습니다!:)

pbg0205 commented on be/src/main/java/codesquad/kr/gyeonggidoidle/issuetracker/domain/milestone/controller/MilestoneController.java:
```diff
@@ -21,4 +31,27 @@ public MilestonePageResponse readOpenMilestones() {
+    }
+
+    @PatchMapping("/api/milestones/{milestoneId}")
+    public ApiResponse updateStatus(@PathVariable Long milestoneId, @RequestParam("isOpen") boolean isOpen) {
```
HTTP Method Patch 인 경우에 query string 을 사용하는 방식이 조금 생소하네요.
흔히 RequestBody 대신 query string 을 사용하신 의도에 대해 질문드려요.

he2joojo replied:
받는 데이터가 ```isOpen```하나라서 RequestBody 대신 query string 을 사용했는데 메서드가 Patch라는걸 깜박했네요! 수정하겠습니다!

he2joojo committed:
[BE]: requestDto 생성자를 public으로 수정 (#79)

fix: requestDto 생성자를 public으로 수정

pbg0205 commented on be/src/main/java/codesquad/kr/gyeonggidoidle/issuetracker/domain/milestone/service/condition/MilestoneCreateCondition.java:
```diff
@@ -0,0 +1,29 @@
+
+    private final String name;
+    private final String description;
+    private final LocalDate dueDate;
```
요청 dueDate 가 과거의 데이터일 저장되나요?

he2joojo replied:
아직 유효성검증을 하지 않아서 과거 데이터일도 저장됩니다. 다음주에 리팩토링하면서 유효성 검증 추가하겠습니다!

pbg0205 commented on be/src/main/java/codesquad/kr/gyeonggidoidle/issuetracker/domain/milestone/service/condition/MilestoneCreateCondition.java:
```diff
@@ -0,0 +1,29 @@
+public class MilestoneCreateCondition {
+
+    private final String name;
+    private final String description;
```
schema 를 확인해보니 `varchar(255)` 로 선언되어 있네요. description 의 경우에는 255자 보다 많은 입력 값이 있을 수가 있어요. 만약 255 자를 넘어갈 경우에 사용자의 데이터가 손실될 수 있어서 이 부분에 대해서도 고민해보시면 좋을 것 같습니다. 설계 시에 고민 요소로 코멘트만 남겨둘게요. (수정은 안하셔도 됩니다.)

he2joojo replied:
db 설계하면서 나중에 수정하자고 했던 부분인데 깜박하고 있었네요! 수정하겠습니다:)

pbg0205 replied:
varchar(2000) 으로 변경해주셨네요 👍👍

pbg0205 commented on be/src/test/java/codesquad/kr/gyeonggidoidle/issuetracker/domain/milestone/integration/MilestoneIntegrationTest.java:
```diff
@@ -1,50 +1,157 @@
+                .build();
+
+        // when
+        Jwt jwt = makeToken();
```
마일스톤을 저장 기능을 테스트하고자 한다면 JWT 를 생성하는 부분은 given 으로 할당해야 할 것 같아요.

he2joojo replied:
수정했습니다:)

pbg0205 commented on be/src/main/java/codesquad/kr/gyeonggidoidle/issuetracker/domain/issue/controller/request/IssueCreateRequest.java:
```diff
@@ -0,0 +1,43 @@
+import java.util.List;
+
+@Getter
+public class IssueCreateRequest {
```
역직렬화를 위해 default constructor 추가해주세요.

he2joojo replied:
수정했습니다:)

pbg0205 commented on be/src/main/java/codesquad/kr/gyeonggidoidle/issuetracker/domain/issue/controller/request/IssueCreateRequest.java:
```diff
@@ -0,0 +1,43 @@
+    private final Long authorId;
+    private final String title;
+    private final String comment;
+    private final List<Long> assignees;
```
만약 assignees 값 중에 null 이 넘어온다면 `null` 로 저장이 되나요?? 만약 DB 에서 NOT NULL 을 선언했다면 DB 에 관련된 예외 메세지가 client 에게 넘어가는지에 대한 여부와 DB 예외가 client 에게 넘어가는 것이 좋은 방법인지에 관한 고민이 필요할 것 같아요.

he2joojo replied:
```
CREATE TABLE issue_label
(
    id       bigint AUTO_INCREMENT,
    issue_id bigint NOT NULL,
    label_id bigint NOT NULL,
    PRIMARY KEY (id)
);
```
db에 NOT NULL로 선언했는데 assignees값이 null이 들어오는 경우 체크를 안해줬네요! IssueService에서 null인지 확인하는 로직을 추가하겠습니다:)

pbg0205 replied:
Service 에서 처리해주셨군요👍👍 validation 을 다룰 수 있는 위치가 다양하게 있어요. 일반적으로 요청에 대한 데이터는 컨트롤러에서 역할을 담당하지만 그렇다면 Service 전용 DTO 를 담당하는 경우에는 어디서 이 값을 확인할 지에 관한 고민도 해보시면 좋을 것 같아요. 현재는 Service 의 비즈니스 로직의 복잡도가 높지 않아 괜찮을 수 있지만 이후에 로직이 비대해지는 경우에는 validation 으로 이한 가독성을 떨어뜨릴 수 있습니다.

he2joojo replied:
validation을 어디서 할지 좀 더 고민해보겠습니다!:)

pbg0205 commented on be/src/main/java/codesquad/kr/gyeonggidoidle/issuetracker/domain/issue/controller/request/IssueStatusRequest.java:
```diff
@@ -0,0 +1,36 @@
+import java.util.List;
+
+@Getter
+public class IssueStatusRequest {
```
역직렬화를 위해 default constructor 추가해주세요.

he2joojo replied:
수정했습니다:)

pbg0205 commented on be/src/main/java/codesquad/kr/gyeonggidoidle/issuetracker/domain/issue/controller/request/IssueUpdateRequest.java:
```diff
@@ -0,0 +1,34 @@
+import java.util.List;
+
+@Getter
+public class IssueUpdateRequest {
```
역직렬화를 위해 default constructor 추가해주세요.

he2joojo replied:
수정했습니다:)

pbg0205 commented on be/src/main/java/codesquad/kr/gyeonggidoidle/issuetracker/domain/issue/controller/response/ApiResponse.java:
```diff
@@ -0,0 +1,25 @@
+import org.springframework.http.HttpStatus;
+
+@Getter
+public class ApiResponse<T> {
```
별도의 ApiResponse 를 선언하신 부분 좋네요👍👍

pbg0205 commented on be/src/main/java/codesquad/kr/gyeonggidoidle/issuetracker/domain/issue/controller/response/ApiResponse.java:
```diff
@@ -0,0 +1,25 @@
+@Getter
+public class ApiResponse<T> {
+
+    private final int statusCode;
```
local test 해보았을 떄 statusCode = 500, status 는 200 OK 를 던지던데,
서버가 예외를 던지더라도 정상 응답을 했을 경우에는 200을 던지는 것으로 협의했나 보군요??

he2joojo replied:
예외처리를 임시로만 해둬서 다음주에 리팩토링하면서 협의할 것 같습니다🥹

pbg0205 replied:
넵 좋습니다👍 만약 협의될 경우에는 IntegrationTest 에서 검증하는 부분에서는 ApiResponse 의 statusCode 를 확인해주시는 로직이 필요해 보여요. http response status 200 인 경우만 추가하는 로직이 종종 있어서 이럴 경우, 로직 검증이 동작하지 않는데 테스트를 통과하는 경우(이를 `거짓 양성` 이라고 합니다.) 가 발생할 수 있어 이 부분도 염두해두시면 좋을 것 같습니다.

pbg0205 commented on be/src/main/java/codesquad/kr/gyeonggidoidle/issuetracker/domain/issue/controller/response/ApiResponse.java:
```diff
@@ -0,0 +1,25 @@
+        return new ApiResponse(status.value(), null);
+    }
+
+    public static ApiResponse exception(HttpStatus status, String data) {
```
윗 메서드가 success 이니까 fail 이라는 메서드 네이밍은 어떤가요?? 
(제안이므로 무조건 수정하실 필요는 없습니다.)

pbg0205 commented on be/src/main/java/codesquad/kr/gyeonggidoidle/issuetracker/domain/issue/controller/IssueController.java:
```diff
@@ -0,0 +1,78 @@
+    @GetMapping("/api/issues/closed")
+    public FilterResponse readCloseIssues() {
+        return FilterResponse.from(issueService.readClosedIssues());
+    }
+    @GetMapping("/api/issues/filtered")
+    public FilterResponse readFiltedIssues(@RequestParam(name = "q", required = false) String encodedQuery) {
```
겸사겸사 개행도 같이 추가해주세요~

pbg0205 commented on be/src/main/java/codesquad/kr/gyeonggidoidle/issuetracker/domain/issue/controller/IssueController.java:
```diff
@@ -0,0 +1,78 @@
+
+    private final IssueService issueService;
+
+    @GetMapping("/api/issues/open")
```
모든 API 에서 prefix 로 사용하는 경우에는 `@RequestMapping("/api/issues")` 사용하는 방법도 있습니다.

pbg0205 commented on be/src/main/java/codesquad/kr/gyeonggidoidle/issuetracker/config/FilterConfig.java:
```diff
@@ -0,0 +1,21 @@
+public class FilterConfig {
+
+    @Bean
+    public FilterRegistrationBean<Filter> jwtAuthorizationFilter(ObjectMapper objectMapper) {
```
Filter 를 적용하는 방법 좋네요👍👍 Filter 와 Interceptor 중에 Filter 를 추가한 이유는 무엇인가요?? 그렇다면 Filter 와 interceptor 의 차이에 대해서도 간단히 학습해보시고 코멘트 남겨주세요.

- [[Spring] 필터(Filter) vs 인터셉터(Interceptor) 차이 및 용도 - (1)](https://mangkyu.tistory.com/173)

pbg0205 commented on be/src/main/java/codesquad/kr/gyeonggidoidle/issuetracker/domain/jwt/JwtAuthorizationFilter.java:
```diff
@@ -0,0 +1,89 @@
+
+public class JwtAuthorizationFilter implements Filter {
+
+    private final String[] whiteListUris = new String[]{"/api/login", "/api/signup", "/api/auth/reissue"};
```
특정 요청에 관한 필터를 활성하는 방법에는 `addPatternUrl()` 메서드를 사용하는 방법도 있어요~

- sample code
```java
    @Bean
    public FilterRegistrationBean<WhaleUrlPatternFilter> uriFilterRegistrationBean() {
        FilterRegistrationBean<WhaleUrlPatternFilter> registrationBean = new FilterRegistrationBean<>();

        registrationBean.setFilter(new WhaleUrlPatternFilter());
        registrationBean.addUrlPatterns("/api/whale/*"); // 이 부분
        registrationBean.setOrder(3); // 호출 순서

        return registrationBean;
    }

```

he2joojo replied:
이런 방법이 있는지 몰랐네요!😃 그런데 ```FilterRegistrationBean```을 사용해서 url을 등록하게 수정하니 통합테스트  몇 개가 실패해서 조금 더 공부하고 적용해보겠습니다!

pbg0205 commented on be/src/main/java/codesquad/kr/gyeonggidoidle/issuetracker/domain/jwt/JwtAuthorizationFilter.java:
```diff
@@ -0,0 +1,89 @@
+
+        if (httpServletRequest.getMethod().equals("OPTIONS")) {
+            return;
+        }
```
매직 스트링보다 spring 에서 제공해주는 변수를 사용해보면 어떨까요??

```java
 if (httpServletRequest.getMethod().equals(HttpMethod.OPTIONS.name())) {
    return;
}

```

pbg0205 commented on be/src/main/java/codesquad/kr/gyeonggidoidle/issuetracker/domain/jwt/JwtAuthorizationFilter.java:
```diff
@@ -0,0 +1,89 @@
+        String authorization = request.getHeader("Authorization");
+        return authorization != null && authorization.startsWith("Bearer ");
+    }
+
```
validation 👍👍👍

pbg0205 commented on be/src/main/java/codesquad/kr/gyeonggidoidle/issuetracker/domain/jwt/JwtAuthorizationFilter.java:
```diff
@@ -0,0 +1,89 @@
+    }
+
+    private boolean isContainToken(HttpServletRequest request) {
+        String authorization = request.getHeader("Authorization");
```
spring 에서 HttpHeaders.AUTHORIZATION 상수를 제공해주고 있어요

pbg0205 commented on be/src/main/java/codesquad/kr/gyeonggidoidle/issuetracker/domain/issue/controller/IssueController.java:
```diff
@@ -0,0 +1,78 @@
+import codesquad.kr.gyeonggidoidle.issuetracker.domain.issue.service.IssueService;
+import lombok.RequiredArgsConstructor;
+import org.springframework.http.HttpStatus;
+import org.springframework.web.bind.annotation.*;
```
(*) 는 인텔리제이 설정을 통해 관리할 수 있습니다

pbg0205 commented on be/src/main/java/codesquad/kr/gyeonggidoidle/issuetracker/domain/issue/service/IssueService.java:
```diff
@@ -29,7 +37,9 @@ public class IssueService {
     private final MemberRepository memberRepository;
+    private final CommentRepository commentRepository;
     private final MilestoneRepository milestoneRepository;
+    private final FilteredIssueRepository filteredIssueRepository;
```
IssueService 가 점점 비대해지는 것 같아요. 비대해지는 클래스는 그만큼 역할과 책임이 커지고 있다는 것을 의미합니다. 한번 고민해볼 필요가 있는 것 같아요.

nag0124 replied:
넵 한번 고려해보겠습니다!

pbg0205 commented on be/src/main/java/codesquad/kr/gyeonggidoidle/issuetracker/domain/issue/Comment.java:
```diff
@@ -0,0 +1,29 @@
+        this.file = file;
+    }
+
+    public boolean isContentsExist() {
+        return this.contents != null;
+    }
```
this.content != null 은 final 이라서 항상 true 를 반환하지 않나요??

he2joojo replied:
놓치고 있던 부분인데 감사합니다🥹 null이 들어올수 있는 변수에 ```final``` 제거하고 ```service```에서 값이 null인지 확인 후에 null이 아닌 경우에만 ```repository``` 호출하도록 수정했습니다!

pbg0205 commented on be/src/main/java/codesquad/kr/gyeonggidoidle/issuetracker/domain/issue/Comment.java:
```diff
@@ -0,0 +1,29 @@
+        return this.contents != null;
+    }
+
+    public boolean isFileExist() {
+        return this.file != null;
+    }
```
이 부분도 this.file != null 은 final 이라서 항상 true 를 반환하는 것 같아요

pbg0205 commented on be/src/main/resources/mapper/filter.xml:
```diff
@@ -0,0 +1,41 @@
+            LEFT JOIN member AS author
+                ON issue.author_id = author.id
+            LEFT JOIN issue_assignee
+                ON issue.id = issue_assignee.issue_id
```
`assignee == null ` 이라면 LEFT JOIN 을 할 필요도 없지 않나요? ? 조회 필드로로 사용되지 않아 불필요해 보여서요.

nag0124 replied:
해당 쿼리에서 label도 비슷하게 JOIN을 합니다.
따라서 조건문으로 감싸 Filter의 필드가 null이 아닌 경우에만 JOIN하도록 변경했습니다.

pbg0205 commented on be/src/main/resources/mapper/filter.xml:
```diff
@@ -0,0 +1,41 @@
+        "http://mybatis.org/dtd/mybatis-3-mapper.dtd">
+<mapper namespace="codesquad.kr.gyeonggidoidle.issuetracker.domain.issue.repository.FilteredIssueRepository">
+    <select id="findByFilter" resultType="IssueVO" parameterType="Filter">
+        SELECT DISTINCT issue.id,
```
DISTINCT 를 선언하셨는데 중복된 데이터가 반환되는 경우가 있었나 보네요??

nag0124 replied:
넵 맞습니다. 테이블을 여러번 조인해서 조회하기 때문에 필터의 조건이 간단한 경우 여러 값들을 보내줍니다.
그런데 저희가 모든 값을 가지고 오는 것이 아니라 작성자 이름, 마일스톤 이름 등 일정 개수의 column만 가지고 오기 때문에 중복이 발생했습니다. 그래서 중복 값을 없애기 위해 DISTINCT를 선언했습니다.

pbg0205 commented on be/src/main/resources/mapper/filter.xml:
```diff
@@ -0,0 +1,41 @@
+        "http://mybatis.org/dtd/mybatis-3-mapper.dtd">
+<mapper namespace="codesquad.kr.gyeonggidoidle.issuetracker.domain.issue.repository.FilteredIssueRepository">
+    <select id="findByFilter" resultType="IssueVO" parameterType="Filter">
+        SELECT DISTINCT issue.id,
```
(참고만 해주세요.) DISTINCT 는 정렬과 그룹핑을 위해 별도의 임시 테이블을 사용하는 방식으로 동작합니다. 만약에 임시 테이블로 만들 데이터가 많을 경우 메모리, 성능상에 문제가 있을 수도 있어요. 저도 직접 경험해본 부분은 아니지만 한번 학습해 볼 포인트가 될 것 같아 코멘트 남깁니다. (keyword: filesort)

nag0124 replied:
넵 학습해보고 적용할 부분이 있으면 하겠습니다.

pbg0205 commented on be/src/main/resources/mapper/filter.xml:
```diff
@@ -0,0 +1,41 @@
+            LEFT JOIN milestone
+                ON issue.milestone_id = milestone.id
+        WHERE issue.is_deleted = false
+          AND issue.is_open = #{isOpen}
```
isOpen 이 `null` 인 경우가 넘어오면 해당되는 조건이 없어서 문제가 있을 수 있을 것 같아요.

nag0124 replied:
먼저 쿼리스트링을 파싱할 때 open이나 close를 넣지 않으면 null을 넣도록 설계했습니다.(Filter 클래스의 checkOpen 참고하시면 됩니다.)
그리고 schema에서 issue Table을 생성할 때 is_open 칼럼을 NOT NULL로 설정하였습니다.
그래서 필터의 isOpen이 Null이면 아무것도 조회하지 못합니다.
이렇게 처리한 이유는는 isOpen의 성질과 관련 있습니다.
isOpen은 모든 필터에 필수적으로 적용되어야 하는 요소라고 생각했습니다.
페이지가 `열린 이슈`와 `닫힌 이슈`로 이미 구분하여 보여주기 때문입니다.
그래서 만약 isOpen에 이상한 값이 들어가 열린 이슈, 닫힌 이슈를 구분하지 않고 가지고 온다면 이는 큰 오류라고 생각했습니다.
이 예외를 처리하기위해 isOpen에 이상한 값이 들어가면 아무것도 가지고 오지 않도록 설계했습니다.
그리고 이를 확인하기 위해 test도 작성했습니다.

pbg0205 commented on be/src/main/java/codesquad/kr/gyeonggidoidle/issuetracker/domain/jwt/controller/request/LoginRequest.java:
```diff
@@ -0,0 +1,29 @@
+
+    @Email
+    private final String email;
+    @NotBlank
```
루카스에 기획서를 확인해보니까 최소 6자리, 최대 16자리로 제약사항이 있어서 요청에 대한 검증 시에 체크가 필요해보입니다.

he2joojo replied:
다음주에 값 유효성 검증하면서 비밀번호 검증도 추가하겠습니다!:)

pbg0205 commented on be/src/main/java/codesquad/kr/gyeonggidoidle/issuetracker/domain/jwt/controller/request/SignUpRequest.java:
```diff
@@ -0,0 +1,33 @@
+import lombok.Getter;
+
+@Getter
+public class SignUpRequest {
```
역직렬화를 위해 기본 생성자 추가해주세요~

pbg0205 commented on be/src/main/java/codesquad/kr/gyeonggidoidle/issuetracker/domain/jwt/controller/request/SignUpRequest.java:
```diff
@@ -0,0 +1,33 @@
+    @Email
+    private final String email;
+    @NotBlank
+    private final String password;
```
루카스 기획서에 보면 password 에 대한 제약사항 추가가 필요해 보여요~ (6<= 길이 <= 16)

pbg0205 replied:
이 부분도 기획서 내용을 반영해주셔야 할 것 같습니다.

pbg0205 commented on be/src/main/java/codesquad/kr/gyeonggidoidle/issuetracker/domain/jwt/JwtAuthorizationFilter.java:
```diff
@@ -0,0 +1,89 @@
+import org.springframework.http.MediaType;
+import org.springframework.util.PatternMatchUtils;
+
+public class JwtAuthorizationFilter implements Filter {
```
token 방식과 session 방식의 중에 token 방식을 선택하신 이유가 무엇인가요?? 만약 token 방식을 사용할 경우의 단점은 없을까요?

he2joojo replied:
세션은 데이터베이스에 저장되지만, 토큰은 클라이언트에서만 저장하기 때문에 stateless해서 확장성 측면에서 유리하다는 장점때문에 세션이 아니라 토큰 방식을 선택했습니다. 반면에 토큰은 노출 가능성이 크고, 서버에 저장하지 않기 때문에 관리하기 어렵다는 단점이 있습니다. 때문에 토큰에는 중요한 정보를 최대한 담지 않고 유효기간을 짧게하고, 로그인 시 유효기간이 긴 refreshToken 추가 발급해 사용합니다:)

pbg0205 replied:
👍👍👍

pbg0205 commented on be/src/main/java/codesquad/kr/gyeonggidoidle/issuetracker/domain/jwt/controller/JwtController.java:
```diff
@@ -0,0 +1,44 @@
+
+@RequiredArgsConstructor
+@RestController
+public class JwtController {
```
JwtController 라는 네이밍이 조금 어색한 것 같아요. 특정 기술에 관한 네이밍 보다는 도메인과 관련된 네이밍이 더 좋지 않을까하는 생각인데 두 분의 의견이 궁금합니다. (저는 Authentication, MemberController 로 선언해서 회원 가입은 MemberController 로 빼고 AuthenticationController 로 생각해봤어요.)

he2joojo replied:
JwtController 네이밍이 조금 어색하고 회원가입 기능이 jwt와 관련 없어서 고민중이었는데 감사합니다:) jwt를 사용하지 않는 회원가입 로직은 ```MemberController```와 ```MemberService```로 분리하고 ```JwtController```는 ```AuthenticationController ```로 네이밍 수정했습니다:)

pbg0205 replied:
수고 많으셨습니다 👍 이후에 패키지도 같이 변경해주시면 좋을 것 같아요
![image](https://github.com/codesquad-members-2023/issue-tracker-max/assets/48561660/7f511ab8-fd29-43b9-b7d9-4004f08f2992)

pbg0205 replied:
JwtRepository 도 네이밍 변경해주세요!

pbg0205 commented on be/src/main/java/codesquad/kr/gyeonggidoidle/issuetracker/domain/jwt/controller/JwtController.java:
```diff
@@ -0,0 +1,44 @@
+    private final JwtService jwtService;
+
+    @PostMapping("/api/login")
+    public JwtResponse login(@RequestBody @Valid LoginRequest request) {
```
@Email, @Password 어노테이션을 통해 만약 유효하지 않는 값이 넘어올 경우 발생하는 예외는 어떤 예외인가요? (해당 예외를 알고 있으면 ExceptionHandler 로 처리하기 편해집니다.)

he2joojo replied:
```MethodArgumentNotValidException```이 발생합니다! ExceptionHandler에 추가했습니다:)

pbg0205 commented on be/src/main/java/codesquad/kr/gyeonggidoidle/issuetracker/domain/jwt/entity/JwtProvider.java:
```diff
@@ -0,0 +1,60 @@
+@Component
+public class JwtProvider {
+
+    private static final byte[] secret = "codesquad-issue-tracker-max-01-gyeonggidoidle".getBytes();
```
일반적으로 secret key 와 같은 값을 properties 로 별도로 빼서 관리합니다. properties 로 관리했을 때의 장점에 대해서 고민해보시면 좋을 것 같습니다.

he2joojo replied:
이 부분도 공부해보고 적용해보겠습니다!

he2joojo replied:
secret key와 토큰 만료시간을 ```application.yml```로 이동시켰습니다!:)

pbg0205 commented on be/src/main/java/codesquad/kr/gyeonggidoidle/issuetracker/domain/jwt/repository/JwtRepository.java:
```diff
@@ -0,0 +1,51 @@
+        try {
+            return template.queryForObject(sql, Map.of("refreshToken", refreshToken), memberRowMapper());
+        } catch (DataAccessException e) {
+            return null;
```
null 을 반환하는 대신 Optional 을 사용해보는건 어떨까요 ?

pbg0205 commented on be/src/main/java/codesquad/kr/gyeonggidoidle/issuetracker/domain/jwt/service/JwtService.java:
```diff
@@ -0,0 +1,72 @@
+    private final MemberRepository memberRepository;
+    private final JwtProvider jwtProvider;
+
+    public Jwt login(LoginCondition condition) {
```
혹시 login 메서드 Transactional 이 실수로 누락하신 건가요? 혹시 의도가 있으시면 말씀 부탁드립니다.

he2joojo replied:
누락했습니다🥹 추가했습니다!

pbg0205 replied:
넵 수고 많으셨습니다👍 만약 해당 클래스의 모든 메서드에 @Transactional 을 사용하는 경우에는 class 상단에 추가해면 중복 코드를 줄여 관리하기 용이해지니 참고해주세요

pbg0205 commented on be/src/main/java/codesquad/kr/gyeonggidoidle/issuetracker/domain/jwt/service/JwtService.java:
```diff
@@ -0,0 +1,72 @@
+        Member member = jwtRepository.findByRefreshToken(refreshToken);
+        if (member == null) {
+            throw new IllegalJwtTokenException(JwtTokenType.REFRESH);
+        }
```
null 을 반환하는 것보다 Optional 을 사용해보는건 어떨까요?

pbg0205 replied:
잘 작성해주셨는데 조금만 첨언을 드리자면 orElseThrow 코드를 줄일 수 있을 것 같아요~
![image](https://github.com/codesquad-members-2023/issue-tracker-max/assets/48561660/8d270538-397f-45fd-9065-caa97969c98a)

pbg0205 commented on be/src/main/java/codesquad/kr/gyeonggidoidle/issuetracker/domain/jwt/service/JwtService.java:
```diff
@@ -0,0 +1,72 @@
+    }
+
+    private boolean verifyPassword(Member member, String password) {
+        return existMember(member) && member.getPassword().equals(password);
```
해당 메서드의 equals member 필드가 null 을 반환할 경우, 예외가 발생할 수 있습니다.`Objects.equals()` 도 있으니 한번 참고해 보세요.

pbg0205 commented on be/src/main/java/codesquad/kr/gyeonggidoidle/issuetracker/domain/label/controller/LabelController.java:
```diff
@@ -16,4 +25,27 @@ public class LabelController {
+    }
+
+    @GetMapping("/api/labels/{labelId}")
+    public LabelDetailsResponse read(@PathVariable Long labelId) {
```
다른 부분은 ApiResponse 인데 LabelDetailsResponse 를 별도로 선언하신 이유가 궁금합니다.

he2joojo replied:
프론트에 넘겨줘야할 데이터가 있는 경우(```GET```)에는 위에와 같이 특정 Response Dto를 반환하고 넘겨줄 데이터가 없는 경우( ```POST, PUT, PATCH, DELETE```)에는 void를 반환하는 대신 ```ApiResponse```를 사용해 상태코드와 메서지를 반환하도록 했습니다. 모든 api에 대해 ```ApiResponse```로 감싸서 반환할까 고민했는데 어떤 방식이 더 좋은지 몰라서 현재 방식으로 선택했습니다! 모든 응답을 ```ApiResponse```로 감싸서 통일성있게 반환하는 것이 더 좋을까요?🥹

pbg0205 replied:
정답은 없지만 일관성을 위해서 `ApResponse` 을 선언하는 것을 선호하는 편이에요

pbg0205 commented on be/src/main/java/codesquad/kr/gyeonggidoidle/issuetracker/domain/member/repository/MemberRepository.java:
```diff
@@ -31,26 +34,79 @@ public Map<Long, List<String>> findAllProfilesByIssueIds(List<Long> issueIds ) {
+        String sql = "SELECT id, email, name, password, profile FROM member WHERE email = :email";
+        try {
+            return template.queryForObject(sql, Map.of("email", email), memberRowMapper());
+        } catch (DataAccessException e) {
+            return null;
+        }
```
예외 발생 시에 null 을 선언하신 의도에 대해 설명 부탁드려요.

he2joojo replied:
예외처리를 다음주에 한번에 하려고 임시로 null로 반환했습니다! 수정하겠습니다:)

pbg0205 commented on be/src/test/java/codesquad/kr/gyeonggidoidle/issuetracker/domain/issue/repository/FilteredIssueRepositoryTest.java:
```diff
@@ -0,0 +1,73 @@
+    @Autowired
+    private SqlSessionFactory sqlSessionFactory;
+
+    @BeforeEach
+    void setUp() {
+        try (SqlSession session = sqlSessionFactory.openSession()) {
+            session.insert("testSchema.createTables");
+            session.insert("testSchema.insertData");
+        }
+    }
+
```
이 코드는 어떤 코드인가요??

nag0124 replied:
본래 마이바티스 테스트 코드를 짜면서 `@Sql`이 동작하지 않는 다고 생각하여 `@MybatisTest`를 통해 `SqlSessionFactory`를 주입하여 `Sqlsession`을 생성하여 `schema.sql`과 `data.sql`을 넣어 DB를 초기화했습니다.
그런데 다시 써보니 되네요...? 수정하겠습니다.

pbg0205 commented on be/src/test/java/codesquad/kr/gyeonggidoidle/issuetracker/domain/issue/service/IssueServiceTest.java:
```diff
@@ -21,101 +21,134 @@
+                    .isEqualTo(1);
+            assertions.assertThat(actual.getMilestoneFilterInformations().get(1).getId())
+                    .isEqualTo(1L);
+        });
```
더미 데이터에 대한 사이즈로 인해서 클래스가 너무 비대해지고 있는 것 같아요. 더미 데이터를 따로 분리해보시는 것도 검토해보시면 좋을 것 같아요.

nag0124 replied:
넵. 하나의 클래스 안에 데이터 생성부분 까지 있으니 가독성이 떨어지는 것 같습니다. 따로 더미데이터를 생성하는 클래스를 만들 수 있으면 만들겠습니다.

pbg0205 commented on be/src/test/java/codesquad/kr/gyeonggidoidle/issuetracker/domain/milestone/repository/MilestoneRepositoryTest.java:
```diff
@@ -1,46 +1,105 @@
+        assertThat(actual).isTrue();
+    }
+
+    @DisplayName("라벨 내용을 수정하고 성공하면 true를 반환한다.")
```
(참고만 해주세요) DisplayName 에 관한 명세의 경우에는 기능 요구사항에 관련해서 명세해주시는 걸 추천드려요. 구체적인 상태를 명세하면 요구사항이 변경 되었을 경우 DisplayName 에 관한 명세도 변경해야 하는 번거로움이 있습니다. 실제 true 인 값은 then절에서 확인이 가능하기도 하구요. 테스트 코드도 유지보수해야 하는 코드라서 이 부분도 한번 고민해보시면 좋을 것 같아요.  저는 아래와 같이 생각해봤습니다. 더 나은 네이밍이나 질문 있으시면 말씀해주세요~

```
@DisplayName("라벨 내용 수정을 성공한다")
```

pbg0205 commented on be/src/test/java/codesquad/kr/gyeonggidoidle/issuetracker/domain/jwt/integration/JwtIntegrationTest.java:
```diff
@@ -0,0 +1,114 @@
+
+        resultActions
+                .andExpect(status().isOk());
+    }
+
+    @DisplayName("refreshToken을 받아서 새로운 accessToken이 담긴 JwtResponse를 반환한다.")
+    @Test
+    @Order(3)
+    @Disabled
+    void reissueAccessToken() throws Exception {
+        Jwt jwt = makeToken();
+        RefreshTokenRequest request = new RefreshTokenRequest(jwt.getRefreshToken());
+        ResultActions resultActions = mockMvc.perform(post("/api/auth/reissue")
+                .contentType(MediaType.APPLICATION_JSON)
+                .content(toJson(request)));
+
+        resultActions
+                .andExpect(status().isOk())
+                .andExpectAll(
+                        jsonPath("$.length()").value(2),
+                        jsonPath("$.accessToken").isNotEmpty(),
```
로그인할 때 토큰을 저장하는데 테스트 코드에는 로그인하는 부분이 없어서 발생하는 이슈인 것 같아요. 한번 확인해보시고 말씀 해주세요~

he2joojo replied:
어디서 문제가 발생하는지 몰라서 한참 헤맸는데 꼼꼼하게 봐주셔서 정말 감사합니다🥹 ```given```절에 로그인 로직을 추가해서 해결했습니다!

pbg0205 commented on be/src/test/java/codesquad/kr/gyeonggidoidle/issuetracker/domain/jwt/integration/JwtIntegrationTest.java:
```diff
@@ -0,0 +1,114 @@
+                        jsonPath("$.length()").value(2),
+                        jsonPath("$.accessToken").isNotEmpty(),
+                        jsonPath("$.refreshToken").isNotEmpty()
+                );
+    }
+
+    @DisplayName("HttpServletRequest에 담긴 claim의 memberId를 가지고 refreshToken을 삭제한다.")
+    @Test
+    @Order(4)
+    void logout() throws Exception {
+        Jwt jwt = makeToken();
+        ResultActions resultActions = mockMvc.perform(post("/api/logout")
+                .header("Authorization", "Bearer " + jwt.getAccessToken()));
```
refresh token 이슈가 원인이 언급드린게 맞다면 해당 부분도 검토가 필요해 보이네요. sample data 로 member2 가 없는 것 같은데 테스트하고 성공을하고 있어서...ㅎㅎ 200 code 외에도  ApiResponse 의 statusCode 검증이 필요해 보여요.

he2joojo replied:
```ApiResponse```의 상태코드 검증도 추가했습니다!

pbg0205 reviewed (changes requested):
안녕하세요. @he2joojo @n-aa-g
이번 주에 많은 작업을 진행하셨네요. 많은 기능 뿐만 아니라 테스트 코드까지 꼼꼼하게 작성해주셨어요👍👍
대단하시면서 피곤하시지 않을까 하는 부분도 염려되네요.
열심히 해주신 덕분에 저도 조금 더 집중하면서 코드 리뷰를 드린 것 같습니다ㅎㅎ 
이번주는 피드백을 좀 많이 드린 것 같은데 읽어보시고 궁금한 점 있으시면 언제든 연락주세요~

nag0124 commented:
첫번째 질문의 핵심은 `핵심 기술만을 사용해서 구현보다는 학습에 가중치를 두는 것이 나을지, 아니면 자동화가 잘 되어있는 기술을 사용하여 학습보다는 구현에 비중을 둘지`였습니다.
이번 주에 기능울 구현하는 과정에서 마이바티스 스프링을 쓴다면 마이바티스 자체를 이해하고 쓰는 것이 아니라고 생각해서 마이바티스만을 사용했었습니다. 다만 마이바티스를 사용하는 과정에서 소위 말하는 `야크 털 깎기`가 많았고 구현을 못할 것 같아 마이바티스-스프링으로 변경했습니다.
그래서 제가 이렇게 핵심 기술을 최소한의 범위로 한정해서 이해하고 사용하려고 하는 방식이 괜찮은 건가 물어보려고 첫번째 질문을 남긴 것이었습니다.
그런데 지금 생각해보니 정답이 없는 문제를 물어본 것 같습니다.
상황에 맞춰서, 시간적 여유가 있거나 학습에 비중을 더 둔다면 핵심기술만을 적용하면서 스스로 연결을 하는 힘든 길을 택할 수도 있고 시간이 촉박하거나 구현에 더 비중을 둔다면 자동화된 도구를 사용하여 먼저 코드를 만들고 이후에 학습할 수도 있다는 생각이 듭니다.
리뷰를 올릴 떄는 제가 정신이 없어 생각이 거기까지 미치지 못했습니다.
이건 개인적인 궁금증인데, 그러면 기본적으로 쿠퍼는 이 두가지 방식중에서 선호하는 방식이 있으신가요? 있으시다면 이유를 알 수 있을까요?

nag0124 committed:
refactor: FilteredIssueRepository 어노테이션 수정

@Sql 을 사용하여 데이터의 주입이 자동화 되게 했습니다.

nag0124 committed:
feat: filter.xml의 동적 쿼리 변경

- Filter에 assginee와 label 필드가 null일 경우 관련 테이블을 join할 필요가 없어 해당 부분의 join을 조건문으로 감쌌습니다.
- Filter의 isOpen이 null인 경우 아무것도 반환하지 않음을 확인하는 테스트를 작성했습니다.

he2joojo committed:
[BE]: 3주차 리뷰 반영 (#81) (#83)

* refactor: requestDto기본 생성자 추가, final 제거 #81

* refactor: 마일스톤 상태 수정 시 @RequestBody로 변경 #81

* refactor: milestone의 description varchar 수정  #81

* refactor: 통합테스트의 makeJwt()를 given절로 이동  #81

* refactor: null이 들어올 수 있는 변수에 final 삭제  #81

* refactor: JwtController를 MemberController와 AuthenticationController로 분리 #81

* refactor: JwtAuthorizationFilter의 매직스트링 제거 #81

* refactor: RepositoryTest @DisplayName 수정 #81

* refactor: JwtIntegrationTest 토큰 재발급, 로그아웃 테스트에 로그인 로직 추가 #81

* refactor: JwtIntegrationTest 로그아웃 상태코드 검증 추가 #81

* refactor: MethodArgumentNotValidException 예외처리 추가 #81

* refactor: 반환타입을 Optional로 수정 #81
